src/Simulation/qdk_sim_rs/src/linalg/expm.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::{
+    fmt::Error,
+    ops::{Mul, Neg},
+    process::Output,
+};
+
+use cauchy::Scalar;
+use ndarray::{
+    linalg::Dot, Array2, ArrayBase, Data, Ix2, OwnedRepr, RawArrayView, RawData, ScalarOperand,
+};
+
+use crate::{
+    error::QdkSimError,
+    linalg::{decompositions::EigenvalueDecomposition, Identity, Inv, MatrixPower},
+    math::approximate_factorial,
+};
+
+use super::{decompositions::ExplicitEigenvalueDecomposition, HasDagger};
+
+// NB: We use the notation of
+//     https://www.cs.jhu.edu/~misha/ReadingSeminar/Papers/Moler03.pdf.
+//
+//     In particular:
+//
+//         e^{A} ≈ R_pq(A) ≔ [D_pq(A)]⁻¹ N_pq(A)
+//
+//     where
+//
+//         N_pq(A) ≔ Σ_j=0^p [(p + q - j)! p!] / [(p + q)! j! (p - j)!] A^j
+//
+//     and
+//
+//         D_pq(A) ≔ Σ_j=0^q [(p + q - j)! p!] / [(p + q)! j! (q - j)!] (-A)^j.
+
+/// Types that support the matrix exponential $e^{A}$.
+pub trait Expm<A>
+where
+    A: Scalar,
+{
+    /// Errors that can result from the [`expm`] method.
+    ///
+    /// [`expm`]: Expm::expm
+    type Error;
+    /// Type used to represent $e^{A}$ when calling [`expm`].
+    ///
+    /// [`expm`]: Expm::expm
+    type Output;
+
+    /// Returns the matrix exponential $e^{A}$ of a matrix $A$.
+    fn expm(&self) -> Result<Self::Output, Self::Error>;
+
+    /// Returns the matrix exponential $e^{\alpha A}$ of a matrix $A$,
+    /// and where $\alpha$ is a complex scalar.
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error>;
+}
+
+impl<A> Expm<A> for ExplicitEigenvalueDecomposition<A>
+where
+    A: Scalar + Mul<Output = A>,
+    ExplicitEigenvalueDecomposition<A>: EigenvalueDecomposition<A>,
+{
+    type Error = <Self as EigenvalueDecomposition<A>>::Error;
+    type Output = <Self as EigenvalueDecomposition<A>>::MatrixSolution;
+
+    fn expm(&self) -> Result<Self::Output, Self::Error> {
+        self.apply_mtx_fn(|x| x.exp())
+    }
+
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error> {
+        self.apply_mtx_fn(|x| (*x * scale).exp())
+    }
+}
+
+#[cfg(feature = "pade")]
+impl<A, S, E> Expm<A> for ArrayBase<S, Ix2>
+where
+    S: Data + RawData<Elem = A>,
+    A: Scalar + ScalarOperand,
+    E: std::convert::From<QdkSimError>,
+    ArrayBase<S, Ix2>: Inv<Error = E> + Identity,
```

Comment:
(Nit, my subjective perception)
I'm personally not a fan of the single-letter identifiers. For template parameters either (I don't know Rust and I assume that `A, S, E` are analogous to the C++'s template parameters). I would recommend to consider more informative identifiers instead of `A, S, E`.
(In C++ it is common to use a single-letter template parameter `T` if it is a data type of the container elements, etc. I'm OK with that. But if there is more than one template param, or it is not a data type (but is an integer constant for example), or is a type with some peculiarities, then the single-letter template params make an impression of sub-optimal code).